third_party/move/move-vm/runtime/src/data_cache.rs

```diff
@@ -146,17 +174,52 @@ impl TransactionDataCache {
         let (layout, contains_delayed_fields) = StorageLayoutConverter::new(module_storage)
             .type_to_type_layout_with_identifier_mappings(ty)?;
 
-        let (data, bytes_loaded) = {
-            let metadata = module_storage
-                .fetch_existing_module_metadata(
-                    &struct_tag.address,
-                    struct_tag.module.as_ident_str(),
-                )
-                .map_err(|err| err.to_partial())?;
-
-            // If we need to process delayed fields, we pass type layout to remote storage. Remote
-            // storage, in turn ensures that all delayed field values are pre-processed.
-            resource_resolver.get_resource_bytes_with_metadata_and_layout(
+        let metadata = module_storage
+            .fetch_existing_module_metadata(
+                &struct_tag.address,
+                struct_tag.module.as_ident_str(),
+            )
+            .map_err(|err| err.to_partial())?;
+
+        let (cached_info, bytes_loaded) = if load_data {
+            let (data, bytes_loaded) = {
+                // If we need to process delayed fields, we pass type layout to remote storage. Remote
+                // storage, in turn ensures that all delayed field values are pre-processed.
+                resource_resolver.get_resource_bytes_with_metadata_and_layout(
+                    addr,
+                    &struct_tag,
+                    &metadata,
+                    if contains_delayed_fields {
+                        Some(&layout)
+                    } else {
+                        None
+                    },
+                )?
+            };
+            let function_value_extension = FunctionValueExtensionAdapter { module_storage };
+            let value = match data {
+                Some(blob) => {
+                    let max_value_nest_depth = function_value_extension.max_value_nest_depth();
+                    let val = ValueSerDeContext::new(max_value_nest_depth)
+                        .with_func_args_deserialization(&function_value_extension)
+                        .with_delayed_fields_serde()
+                        .deserialize(&blob, &layout)
+                        .ok_or_else(|| {
+                            let msg = format!(
+                                "Failed to deserialize resource {} at {}!",
+                                struct_tag.to_canonical_string(),
+                                addr
+                            );
+                            PartialVMError::new(StatusCode::FAILED_TO_DESERIALIZE_RESOURCE)
+                                .with_message(msg)
+                        })?;
+                    GlobalValue::cached(val)?
+                },
+                None => GlobalValue::none(),
+            };
+            (CachedInformation::Value(value), bytes_loaded)
```

Comment:
nit: probably cast these `bytes_loaded` to `u64`, to avoid cast below and later when doing `NumBytes::new(bytes_loaded as u64)`